src/main/java/net/minecraftforge/network/simple/IndexedMessageCodec.java

```diff
@@ -150,6 +150,7 @@ MSG newInstance() {
     void consume(FriendlyByteBuf payload, int payloadIndex, Supplier<NetworkEvent.Context> context) {
         if (payload == null) {
             LOGGER.error(SIMPLENET, "Received empty payload on channel {}", Optional.ofNullable(networkInstance).map(NetworkInstance::getChannelName).map(Objects::toString).orElse("MISSING CHANNEL"));
+            context.get().setPacketHandled(true); //don't disconnect because of empty packets
```

Comment:
This should only be done for packets that we actually know of and that do not expect a response.

Reply:
Because the payload is null at this point, it's difficult to look up what packet on the client caused this empty-payload-packet to be sent to the server. I thought logging an error about an empty payload packet being received is sufficient and that the server should just move on with the handshake if the client isn't able to decode a packet (which currently only happens to `S2CModData` anyways, a packet that the server doesn't care about if the client was able to decode it or not). If a packet that is not declared as not needing a response (and thus, for that packet, it is crucial for a response of this packet to be sent) can't be read by the client, the empty payload packet sent because of it will be ignored/not disconnect at this point in the code, but the handshake will fail anyways because the needed response never arrives (see HandshakeHandler#tickServer).

Reply:
Can't you look it up based on the login index?

Reply:
Just thought about it, I could maybe test if the given `payloadIndex` is contained in `sentMessages` (if it is, the packet requires a response, and if not, it doesn't because of the filter in `HandshakeHandler#tickServer`). Should I try to do this (and then not mark the packet as handled if an actual response is required for the packet with that login index)?

Reply:
If it's possible without too much hackery, sure. If not, it should be OK because of the other thing you mentioned (I will have to test)

Reply:
I now added a check as described in my comment above at the desired place, hopefully it isn't considered too hacky (I at least don't think so, but if you have another opinion feel free to tell me)